### DISCUSSION TITLE
Is gleam a subset of Python?

### DISCUSSION THREAD

**Discussion by pwtail** [gleam-lang/gleam#2057] [Ideas & suggestions]
Hi @lpil

I happen to be a Python developer, and I think noticed a thing: is gleam a subset of Python? Python has added the match..case statement in 3.10 for pattern matching, with types, guards and everything.

Of course, Python is not a functional language, so `if`, `for` and `match` are not expressions, but simply blocks. But other than that?

Yes, I know, Jose already used the Ruby syntax. But he copied the OOP part, which he shouldn't!  By no means! :)

**Comment by lpil** [comment]
Hi @pwtail ! No, Gleam is not a Python subset, it is very different in many ways. I would be happy to go into any details you may be interested in.

**Comment by pwtail** [comment]
If you don't mind... then what are the parts missing from Python?

I have found one thing myself, that is the pipeline operator. Will you find
more?)



ср, 8 мар. 2023 г., 15:48 Louis Pilfold ***@***.***>:

**Reply by lpil** [reply]
Other language features include the static type system, compile time resolution of function argument labels, tail call optimisation, different semantics for pattern matching, `use` expressions, and a very different syntax that will allow things like multi-statement lambda functions. If you look at the runtime, the tooling, and the ecosystem there are further differences too, and there are many features that Python has that Gleam deliberately does not.

**Comment by pwtail** [comment]
>and there are many features that Python has that Gleam deliberately does not.

That's why I used the word "subset". The runtime is different too, there is
no doubt about it.

I've read somewhere the inspiration for the "use" keyword was the python's
'with', so I didn't count it.

ср, 8 мар. 2023 г., 16:03 Louis Pilfold ***@***.***>:

**Reply by lpil** [reply]
While `with` was one of the influences `use` is very different and more flexible. If anything one of the things that inspired `use` was my frustrations with how limited `with` is in Python 😁

**Comment by pwtail** [comment]
Sorry I feel I'm being very rude right now. The first thing is that gleam
does things just right and I'm very impressed.

ср, 8 мар. 2023 г., 16:12 Абеткин Виталий ***@***.***>:

**Comment by pwtail** [comment]
Python has decorators, on the other hand)

ср, 8 мар. 2023 г., 16:41 Louis Pilfold ***@***.***>:

**Reply by lpil** [reply]
Python's decorators are also very limited, being just another syntax for function calls. In Gleam you can achieve the same with `use`.

**Reply by pwtail** [reply]
Yes they are. However it is tempting to make a gleam remake with python syntax. Please tell me something to rethink this :)

**Reply by lpil** [reply]
Sounds like it could be a fun project!

**Reply by pwtail** [reply]
Or it will be useless. If there are no new ideas to bring in, why start it?

Soon there will be the necessity to break the python syntax (to let use pattern matching inside the `if` statement, for example). So, soon it will be invalid python code anyway. So is the gleam code currently.
